### DISCUSSION TITLE
Question about per-host runtime determinism

### DISCUSSION THREAD

**Discussion by mlowicki** [tokio-rs/turmoil#195] [General]
Hi there,

I’ve read a bit of the code to understand how things work under the hood and have question about determinism. Every host uses its own tokio’s single-threaded executor and Sim simply iterates over hosts in a deterministic order and ticks every host. Is determinism of per-host runtime guaranteed somehow? In other words how it’s achieved that host has some tasks to run and execution of those will be deterministic?

**Comment by mcches** [comment]
Hi, thanks for taking a look at turmoil! Within each host we use a paused runtime that we step forward deterministically with each step using the LocalSet. Our use of the current thread, LocalSet, emulated sockets, and this timing mechanism make all of this work. 

Here is where all the magic happens: https://github.com/tokio-rs/turmoil/blob/main/src/rt.rs#L126-L148

That said, its up to the software (just a future) that runs on each host to do the rest. Things like iterating a `HashMap` or tokio's `select!` can introduce non-determinism. We've looked at this before, but its pretty difficult to cover everything out there.

**Comment by mcches** [comment]
We've got a long standing issue to create some documentation about this: https://github.com/tokio-rs/turmoil/issues/19. It just hasn't been picked up yet.

**Reply by mlowicki** [reply]
Thanks! I’m planning to deep dive into how everything works so might be able to tackle #19 at some point.

If you don’t mind I’ll use this thread for additional questions.

Do I understand right that `LocalSet` contains only software and when running `tick()` there is additionally sleep future to make progress with time limit? Since we’ve only those in LocalSet then as long as software is deterministic the whole `tick()` should be deterministic?

**Reply by mcches** [reply]
Yes, that's correct. We basically say poll tasks in the provided software for the tick duration and because time is paused we achieve determinism here.

**Comment by mlowicki** [comment]
@mcches 👋 can you also explain a bit how the timing mechanism works?

I've a simulation over some network calls and I use timeouts via `tokio::time::timeout`. When I do `turmoil::hold("broker", "client");` before issuing a call with 10s timeout then reach timeout obviously. The test runs in 0.03s but the simulated time is many seconds (I see simulation duration to 30s). Wondering how this is achieved? I see the timer of a current host is being updated during each step (https://github.com/tokio-rs/turmoil/blob/main/src/sim.rs#L412) but would like to understand how that is then glued with my call to  `tokio::time::timeout(Duration::from_secs(10)..`.

**Comment by mcches** [comment]
All of this is managed by tokio. Each host has a tokio runtime, which is created here: https://github.com/tokio-rs/turmoil/blob/main/src/rt.rs#L227-L231. Note `start_paused(true)`.

ref: https://docs.rs/tokio/latest/tokio/runtime/struct.Builder.html#method.start_paused

Each step() of the simulation moves the runtime forward `tick_duration` via a `LocalSet` in the host: https://github.com/tokio-rs/turmoil/blob/main/src/rt.rs#L149-L155. Tokio's `time::` constructs "just work", we aren't doing anything special here.

**Reply by mlowicki** [reply]
default value for `tick_duration` is 1ms so tick of each runtime means:

```
    pub(crate) fn tick(&mut self, duration: Duration) -> Result<bool> {
        self.tokio.block_on(async {
            self.local
                .run_until(async {
                    sleep(1ms).await;
                })
                .await
        });
```

but this would be waiting for 1ms as seen by wall clock time (real, not simulated time), right? So still not clear to my how my `tokio::time::timeout(Duration::from_secs(10)..` timeouts out even if on wall clock time only 0.03s elapsed 🤔 .

**Comment by mcches** [comment]
It's not 1ms wall clock when the runtime is paused, its essentially instantaneous. The tokio docs are pretty thorough about how this all works: https://docs.rs/tokio/latest/tokio/time/fn.pause.html.

**Reply by mlowicki** [reply]
> It's not 1ms wall clock when the runtime is paused, its essentially instantaneous. The tokio docs are pretty thorough about how this all works: https://docs.rs/tokio/latest/tokio/time/fn.pause.html.

Thanks a lot. Now it's way clearer.
